#+PROPERTY: header-args:sh :results output :exports both :shebang "#!/bin/bash"
#+PROPERTY: header-args:R :session "* R-kelsmith-salmonella *" :results output :exports results
#+OPTIONS: ^:nil
#+LATEX_HEADER: \usepackage[margin=0.5in]{geometry}
#+TITLE: Kelly Smith Salmonella project data exploration
#+AUTHOR: Noah Hoffman

* Overview of methods

- paired-end Illumina sequencing (V4 variable region, PCR primers 515/806) performed by MrDNA[fn:: www.mrdnalab.com]
- self-assembly, primer trimming performed by MrDNA
- Reads from all specimens were pooled and clustered with swarm with d=1[fn:: https://github.com/torognes/swarm] using swarmwrapper[fn:: https://github.com/nhoffman/swarmwrapper] to discard OTUs of size 1.
- A phylogenetic reference tree was created by recruiting 16S rRNA reference sequences from a curated subset of RDP based on similarity to cluster centroids.
- Taxonomic assignment was performed by placing aligned cluster centroids onto the reference tree with pplacer
- Length PCA was performed using =guppy lpca=[fn:: https://matsen.github.io/pplacer/generated_rst/guppy_lpca.html#guppy-lpca]

* Sample annotation

#+BEGIN_SRC R
labels <- read.csv('../2016-05-16_kelsmith_data_prep/output/salmonella/labels.csv',
                   colClasses=c(specimen='character'))

pca_data <- read.csv('output/lpca.proj', header=FALSE)
colnames(pca_data) <- c('specimen', gettextf('pc%s', seq(ncol(pca_data) - 1)))
pca_data$specimen <- as.character(pca_data$specimen)
tab <- merge(pca_data, labels, by='specimen', all.x=FALSE, all.y=TRUE)
tab <- within(tab, {
  antibiotic <- factor(ifelse(antibiotic == '', 'none', as.character(antibiotic)))
})
tab$day <- factor(gettextf('day_%i', tab$day))
summary(tab[,-1*seq(6)])
#+END_SRC

#+RESULTS:
#+begin_example
 sex          strain      day            antibiotic salmonella   cecum_cfu
 F:30   B6       :26   day_0:66   ampicilin   :22   no :66     Min.   : 0.000
 M:67   Naip5    :25   day_1:31   none        :19   yes:31     1st Qu.: 0.000
        Rag1     :18              streptomycin:56              Median : 0.000
        red_Naip5:28                                           Mean   : 2.316
                                                               3rd Qu.: 5.885
                                                               Max.   :10.204
 trichomonas e_coli
 no :54      no :94
 yes:43      yes: 3
#+end_example

* length PCA

** PCA plots annotated with sample metadata

Unlabeled plot of first two principal components.

#+BEGIN_SRC sh
mkdir -p figs
#+END_SRC

#+BEGIN_SRC R :results output graphics :file figs/lpca1.pdf
library(lattice, quietly = TRUE)
library(latticeExtra, quietly = TRUE)
ff <- xyplot(pc2 ~ pc1,
             data=tab,
             par.settings=theEconomist.theme(),
             auto.key=list(space='right'),
             main='Length PCA',
             sub='no features annotated'
             )
plot(ff)
#+END_SRC

#+BEGIN_SRC R :results output graphics :file figs/lpca7.pdf
ff <- xyplot(pc2 ~ pc1 | salmonella + antibiotic,
             groups=strain,
             data=tab,
             par.settings=theEconomist.theme(),
             auto.key=list(space='right'),
             main='Length PCA',
             sub='partition by salmonella inoculation and antibiotic treatment, group by strain'
             )
plot(ff)
#+END_SRC

#+RESULTS:
[[file:figs/lpca7.pdf]]

** PCA plots annotated by bacterial composition

Order or Family-level classifications are used to annotate points representing specimens in the PCA plots below.

#+NAME:   fig:pies_order
#+CAPTION: Length PCA with order-level classifications
file:output/pies_order.pdf

#+NAME:   fig:pies_family
#+CAPTION: Length PCA with family-level classifications
[[file:output/pies_family.pdf]]

* version information

#+BEGIN_SRC sh
pwd
git --no-pager log -n1
git --no-pager status
git remote -v
#+END_SRC

#+RESULTS:
#+begin_example
/fh/fast/fredricks_d/bvdiversity/working/nhoffman/2016-05-31_kelsmith_salmonella
commit 82870fa075661d91c07cd8a370add67de14ab245
Author: Noah Hoffman <noah.hoffman@gmail.com>
Date:   Tue Jun 21 10:57:15 2016 -0700

    add overview of methods
On branch 2016-05-31_kelsmith_salmonella
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   2016-05-31_kelsmith_salmonella.org

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	2016-05-31_kelsmith_salmonella.html
	2016-05-31_kelsmith_salmonella.pdf
	2016-05-31_kelsmith_salmonella.t2d/
	2016-05-31_kelsmith_salmonella.tex
	figs/

no changes added to commit (use "git add" and/or "git commit -a")
origin	git@github.com:fhcrc/yapp.git (fetch)
origin	git@github.com:fhcrc/yapp.git (push)
#+end_example
